events_test/src/main/java/com/adobe/aio/event/management/ProviderServiceTester.java

```diff
@@ -0,0 +1,116 @@
+        .description(label + DEFAULT_DESC_SUFFIX)
+        .docsUrl(DEFAULT_PROVIDER_DOCS_URL);
+  }
+
+  public Provider createOrUpdateProvider(String providerLabel, String eventCode) {
+    return createOrUpdateProvider(getTestProviderInputModelBuilder(providerLabel).build(),
+        Set.of(getTestEventMetadataBuilder(eventCode).build()));
+  }
+
+  public Provider createOrUpdateProvider(ProviderInputModel providerInputModel,
+      Set<EventMetadata> eventMetadataSet) {
+    return assertProviderResponseAndCreateEventMetadata(providerInputModel, eventMetadataSet,
+        () -> providerService.createOrUpdateProvider(providerInputModel));
+  }
+
+  public Provider createProvider(ProviderInputModel providerInputModel,
+      Set<EventMetadata> eventMetadataSet) {
+    return assertProviderResponseAndCreateEventMetadata(providerInputModel, eventMetadataSet,
+        () -> providerService.createProvider(providerInputModel));
+  }
+
+  private Provider assertProviderResponseAndCreateEventMetadata(ProviderInputModel providerInputModel,
+      Set<EventMetadata> eventMetadataSet, Supplier<Optional<Provider>> providerSupplier) {
+    Optional<Provider> provider = providerSupplier.get();
+    Assert.assertTrue(provider.isPresent());
+    logger.info("Created AIO Events Provider: {}", provider);
+    String providerId = provider.get().getId();
+    Assert.assertTrue(StringUtils.isNotBlank(providerId));
+    Assert.assertTrue(StringUtils.isNotBlank(provider.get().getInstanceId()));
+    Assert.assertNotNull(provider.get().getId());
+    Assert.assertNotNull(provider.get().getInstanceId());
+    Assert.assertEquals(providerInputModel.getLabel(), provider.get().getLabel());
+    Assert.assertEquals(providerInputModel.getDescription(), provider.get().getDescription());
+    Assert.assertEquals(providerInputModel.getDocsUrl(), provider.get().getDocsUrl());
+    Assert.assertEquals(WorkspaceUtil.getSystemProperty(Workspace.IMS_ORG_ID),
+        provider.get().getPublisher());
+    Assert.assertEquals(CloudEvent.SOURCE_URN_PREFIX + provider.get().getId(),
+        provider.get().getSource());
+    Assert.assertEquals(providerInputModel.getProviderMetadataId(),
+        provider.get().getProviderMetadata());
+
+    for (EventMetadata eventMetadataInput : eventMetadataSet) {
+      Optional<EventMetadata> eventMetadata = providerService.createEventMetadata(providerId,
+          eventMetadataInput);
+      Assert.assertTrue(eventMetadata.isPresent());
+      Assert.assertEquals(eventMetadataInput.getEventCode(), eventMetadata.get().getEventCode());
+      Assert.assertEquals(eventMetadataInput.getDescription(),
+          eventMetadata.get().getDescription());
+      Assert.assertEquals(eventMetadataInput.getEventCode(), eventMetadata.get().getLabel());
+      logger.info("Added EventMetadata `{}` to AIO Events Provider `{}`", eventMetadata,
+          providerId);
+    }
+    return provider.get();
+  }
+
+
+  public void deleteProvider(String providerId) {
```

Comment:
many times i have seen that when an assertion fails, the test stops and the provider that is created is not cleaned up. Then we end up with a lot of test providers in that org. Can we ensure that the cleanup happens even if the test fails?

Reply:
same for other tests

Reply:
thanks. done